# 06 - Stride Ablation, Overlapping Training Windows

All models in notebooks 01–05 were trained on **non overlapping** windows:
each window starts exactly one window length after the previous one.
At ~80 rows/s this yields roughly 700 training windows for W3 (5 s) and 350
for W4 (10 s) potentially insufficient for data hungry models like CNN and
Transformer.

This notebook asks a single question:

> **Does sliding the training window with overlap (smaller stride) improve
> classification performance, and does the benefit vary across model families?**

The **test split is always non overlapping** (stride = window size) to ensure
a fair, leak free evaluation. Only the *training* stride is varied.

### Approach

1. **Stride sweep on LR** (fast proxy) AUC averaged over 3 seeds to find the
   best stride per window size without overfitting to a single run.
2. **All five models** reevaluated at their window's best stride with the same
   multi seed protocol used in notebooks 01–05.
3. **ΔF1 / ΔAUC** reported against the non overlapping baseline from each model's
   dedicated notebook.

### Key findings

| Finding | Detail |
|---------|--------|
| Short windows benefit most | W2 (1 s) improves across **all** models |
| LR recovers dramatically | W4 F1: 0.791 => **0.951** (+0.160) was sample starved |
| CNN biggest single gain | W2 F1: 0.826 => **0.961** (+0.136) |
| OCSVM degrades at W3/W4 | Near duplicate training windows crowd the one class boundary |
| Overlap is not universally good | Best stride is a per model, per window hyperparameter |

## 1 - Imports & Configuration

All hyperparameters are declared once here so every subsequent cell can be
rerun by changing a single value.

**Window sizes** correspond to physical durations at the ESP32 sampling rate of
~80 rows/s: W2 ≈ 1 s, W3 ≈ 5 s, W4 ≈ 10 s.

**Stride grids** are defined per window size. The smallest stride tested is
0.1 s for W2 (8 rows) and 0.5 s for W3/W4 (40 rows), giving up to ~92 %
overlap between consecutive training windows.

**STRIDE_LABELS** is a nested dict keyed by `(window, stride_rows)` rather
than a flat dict stride=40 rows means 0.5 s for W2 but a different fraction
for W3/W4, so a flat key would be ambiguous in comparison plots.

**Device** follows CUDA => MPS => CPU priority so the notebook runs unchanged
on a cloud GPU, Apple Silicon, or a CPU only machine.

**BASELINE** stores the seed-42 F1 and AUC from notebooks 01–05 (non overlapping
windows) as the reference point for all ΔF1 / ΔAUC comparisons below.

In [ ]:
import sys, math, random, importlib, os
import numpy as np
import pandas as pd
import torch
import torch.nn as nn
from torch.utils.data import DataLoader, TensorDataset
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from sklearn.linear_model import LogisticRegression
from sklearn.svm import OneClassSVM
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score, roc_auc_score, roc_curve

import utils
importlib.reload(utils)
from utils import FEATURE_COLS, TEST_SESSIONS, load_data

SAVE_DIR = "full-study-outputs"
os.makedirs(SAVE_DIR, exist_ok=True)

# --- Reproducibility ---------------------------------------------------------
SEED = 42
random.seed(SEED)
np.random.seed(SEED)
torch.manual_seed(SEED)
torch.cuda.manual_seed_all(SEED)

# --- Device ------------------------------------------------------------------
DEVICE = (
    torch.device("cuda") if torch.cuda.is_available() else
    torch.device("mps")  if torch.backends.mps.is_available() else
    torch.device("cpu")
)
print(f"Device : {DEVICE}")

# --- Window sizes ------------------------------------------------------------
WSIZE = {"W2_1s": 80, "W3_5s": 400, "W4_10s": 800}

STRIDES = {
    "W2_1s":  [80, 40, 20, 10, 8],
    "W3_5s":  [400, 320, 160, 80, 40],
    "W4_10s": [800, 560, 320, 160, 80, 40],
}

STRIDE_LABELS = {
    "W2_1s":  {80: "1s (base)", 40: "0.5s", 20: "0.25s", 10: "0.125s", 8: "0.1s"},
    "W3_5s":  {400: "5s (base)", 320: "4s", 160: "2s", 80: "1s", 40: "0.5s"},
    "W4_10s": {800: "10s (base)", 560: "7s", 320: "4s", 160: "2s", 80: "1s", 40: "0.5s"},
}

N_FEATURES = 55

# --- Hyper-parameters --------------------------------------------------------
EPOCHS_AE    = 60;   BATCH_AE  = 512;  LR_AE       = 1e-3;  AE_VAL_FRAC = 0.15
EPOCHS_CNN   = 80;   LR_CNN    = 1e-3; CNN_PATIENCE = 15
BATCH_CNN    = {"W2_1s": 256, "W3_5s": 128, "W4_10s": 64}
EPOCHS_TR    = 100;  LR_TR     = 1e-3; TR_PATIENCE  = 15
BATCH_TR     = {"W2_1s": 256, "W3_5s": 64,  "W4_10s": 64}

SEEDS = [42, 0, 123]

# --- Non-overlapping baselines -----------------------------------------------
BASELINE = {
    "LR": {
        "W2_1s":  {"f1": 0.8559, "auc": 0.9107},
        "W3_5s":  {"f1": 0.8030, "auc": 0.8149},
        "W4_10s": {"f1": 0.7907, "auc": 0.8038},
    },
    "OCSVM": {
        "W2_1s":  {"f1": 0.7819, "auc": 0.9184},
        "W3_5s":  {"f1": 0.9408, "auc": 0.9763},
        "W4_10s": {"f1": 0.9536, "auc": 0.9935},
    },
    "AE": {
        "W2_1s":  {"f1": 0.9510, "auc": 0.9821},
        "W3_5s":  {"f1": 0.9259, "auc": 0.9849},
        "W4_10s": {"f1": 0.9429, "auc": 0.9562},
    },
    "CNN": {
        "W2_1s":  {"f1": 0.8256, "auc": 0.9685},
        "W3_5s":  {"f1": 0.8114, "auc": 0.9543},
        "W4_10s": {"f1": 0.8092, "auc": 0.9332},
    },
    "Transformer": {
        "W2_1s":  {"f1": 0.9132, "auc": 0.9571},
        "W3_5s":  {"f1": 0.8060, "auc": 0.9156},
        "W4_10s": {"f1": 0.7977, "auc": 0.9414},
    },
}

results = {wk: {} for wk in WSIZE}

print(f"EPOCHS_CNN={EPOCHS_CNN}  CNN_PATIENCE={CNN_PATIENCE}")
print(f"EPOCHS_TR={EPOCHS_TR}   TR_PATIENCE={TR_PATIENCE}")
print(f"BATCH_AE={BATCH_AE}  BATCH_CNN={BATCH_CNN}  BATCH_TR={BATCH_TR}")

Device : cuda
EPOCHS_CNN=80  CNN_PATIENCE=15
EPOCHS_TR=100   TR_PATIENCE=15
BATCH_AE=512  BATCH_CNN={'W2_1s': 256, 'W3_5s': 128, 'W4_10s': 64}  BATCH_TR={'W2_1s': 256, 'W3_5s': 64, 'W4_10s': 64}


CUDA is available all neural models (AE, CNN, Transformer) run on GPU.
Batch sizes scale down with window size: larger windows consume more memory
per sample, so W4 uses 4× smaller batches than W2 to keep GPU utilisation
stable. Early stopping patience of 15 epochs gives models enough room to
recover from temporary plateaus before terminating.

## 2 - Data Loading

Raw CSI frames are loaded from the shared preprocessed file produced in
`00_preprocessing.ipynb`. Each row is one ESP32 CSI measurement: 51 subcarrier
amplitudes + 4 scalar statistics = **55 features**, plus a `session` identifier
and a binary `label` (empty / occupied).

The `session` column is critical all windowing functions group by session
first so windows never cross a recording boundary.

In [3]:
df = load_data()

print(f"Total rows  : {len(df):,}")
print(f"Sessions    : {df['session'].nunique()}")
print(f"Labels      : {df['label'].value_counts().to_dict()}")
print(f"Test sessions: {TEST_SESSIONS}")
print()
print(df.head(3))

Total rows  : 376,514
Sessions    : 11
Labels      : {'empty': 189933, 'occupied': 186581}
Test sessions: ['session5_morning_empty', 'session7_morning_occupied']

                         time  label  rssi  noise   amp_2   amp_3   amp_4  \
0  2026-09-28T21:40:35.385507  empty   -53    -97  21.000  16.279  13.416   
1  2026-09-28T21:40:35.416333  empty   -53    -97  20.248  18.248  18.000   
2  2026-09-28T21:40:35.435720  empty   -53    -97  21.633  20.518  21.633   

    amp_5   amp_6   amp_7  ...  amp_60  amp_61  amp_62  \
0  14.142  15.811  19.235  ...  35.355  36.125  36.014   
1  18.248  20.616  21.840  ...  32.249  31.906  31.064   
2  23.707  25.942  28.178  ...  35.903  34.713  32.558   

                       session     day   csi_mean   csi_std  csi_max  \
0  01_day_session1_night_empty  01_day  20.226294  8.443706   36.125   
1  01_day_session1_night_empty  01_day  20.399098  8.110109   32.650   
2  01_day_session1_night_empty  01_day  23.605118  9.434751   37.363   

     c

376,514 frames across 11 sessions nearly balanced at 50.5 % empty /
49.5 % occupied, so no class weighting or resampling is needed.

Two sessions are held out for evaluation: one empty (`session5_morning_empty`)
and one occupied (`session7_morning_occupied`), both recorded on a different day
from the training sessions. This cross day split is the hardest realistic test:
the model must generalise to a new day, not just a new time slot.

## 3 - Windowing Function

`make_windows_strided` slides a fixed length window over each session
independently windows never cross a recording boundary. The training split
uses the requested stride, the test split always uses stride = window size
(non overlapping) so evaluation windows are independent and the test set size
is fixed across all stride experiments.

`aggregate_window` collapses the time axis by concatenating per feature mean
and standard deviation: `(N, T, 55) => (N, 110)`. This flattened representation
is used by LR and OCSVM, CNN and Transformer consume the raw `(N, T, 55)` tensor.

Window labels are assigned by **majority vote**: a window is labelled *occupied*
if more than 50 % of its frames are occupied.

In [ ]:
def _is_test_session(session_series):
    """Boolean mask: True for rows belonging to test sessions"""
    mask = pd.Series(False, index=session_series.index)
    for s in TEST_SESSIONS:
        mask |= session_series.str.contains(s, regex=False)
    return mask


def make_windows_strided(df, wsize, stride):
    """
    Overlapping stride windowing for training non overlapping for test.

    FIX: windows generated PER SESSION before concatenation no window
    ever spans rows from two different recordings.

    Returns X_tr (N,T,55), y_tr (N,), X_te (M,T,55), y_te (M,)
    """
    test_mask = _is_test_session(df["session"])
    train_df  = df[~test_mask]
    test_df   = df[test_mask]

    def _slide(sub_df, step):
        Xl, yl = [], []
        for _, grp in sub_df.groupby("session", sort=False):
            feats  = grp[FEATURE_COLS].values.astype(np.float32)
            labels = (grp["label"] == "occupied").astype(int).values
            for s in range(0, len(feats) - wsize + 1, step):
                Xl.append(feats[s:s + wsize])
                yl.append(int(labels[s:s + wsize].mean() >= 0.5))
        return (np.array(Xl, dtype=np.float32),
                np.array(yl, dtype=np.int64))

    X_tr, y_tr = _slide(train_df, stride)
    X_te, y_te = _slide(test_df,  wsize)   # test always non overlapping
    return X_tr, y_tr, X_te, y_te


def aggregate_window(X):
    """(N, T, 55) => (N, 110)  via mean + std across the time axis"""
    return np.concatenate([X.mean(axis=1), X.std(axis=1)], axis=1)


# --- Preview window counts ---------------------------------------------
print(f"{'Window':<18}  {'Stride':>12}  {'Train':>8}  {'Test':>6}")
print("-" * 52)
for wk, wsize in WSIZE.items():
    print()
    for stride in STRIDES[wk]:
        Xtr, _, Xte, _ = make_windows_strided(df, wsize, stride)
        label = STRIDE_LABELS[wk][stride]
        print(f"  {wk:<18}  {label:>12}  {len(Xtr):>8,}  {len(Xte):>6,}")

Window                    Stride     Train    Test
----------------------------------------------------

  W2_1s                  1s (base)     3,650   1,051
  W2_1s                       0.5s     7,295   1,051
  W2_1s                      0.25s    14,587   1,051
  W2_1s                     0.125s    29,170   1,051
  W2_1s                       0.1s    36,460   1,051

  W3_5s                  5s (base)       726     209
  W3_5s                         4s       907     209
  W3_5s                         2s     1,809     209
  W3_5s                         1s     3,614     209
  W3_5s                       0.5s     7,223     209

  W4_10s                10s (base)       360     104
  W4_10s                        7s       512     104
  W4_10s                        4s       895     104
  W4_10s                        2s     1,787     104
  W4_10s                        1s     3,569     104
  W4_10s                      0.5s     7,133     104


The table shows the core trade-off of overlapping windows: training set size
scales almost linearly as stride shrinks, while the test set stays fixed.

At W4 (10 s), the non overlapping baseline produces only **360 training
windows** too few for CNN and Transformer to learn reliably. The 0.5 s stride
expands that to **7,133** (20× more), which explains the large ΔF1 gains seen
later for those models.

The test set is intentionally small (104–1,051 windows) because it covers only
two held out sessions. This is realistic for a deployed sensor: you evaluate on
what you actually collected, not a synthetic split.

## 4 - LR Stride Sweep

Logistic Regression is used as a fast, cheap proxy to find the best stride
per window size. Running the full sweep on CNN or Transformer would take
hours LR completes in seconds while still capturing the data quantity effect
that overlapping windows produce.

Each stride is evaluated by averaging AUC over **3 seeds** (42, 0, 123) to
reduce variance from random weight initialisation. The stride with the highest
mean AUC is then carried forward for all five models in the cells below.

Features are aggregated before fitting: `(N, T, 55) => (N, 110)` via
`aggregate_window`, then standardised with `StandardScaler` fit on training
data only.

In [5]:
def _run_lr(Xtr_agg, y_tr, Xte_agg, y_te, seed):
    scaler  = StandardScaler()
    Xtr_sc  = scaler.fit_transform(Xtr_agg)
    Xte_sc  = scaler.transform(Xte_agg)
    clf     = LogisticRegression(max_iter=1000, C=1.0,
                                  class_weight="balanced",
                                  random_state=seed, n_jobs=-1)
    clf.fit(Xtr_sc, y_tr)
    probs = clf.predict_proba(Xte_sc)[:, 1]
    return f1_score(y_te, clf.predict(Xte_sc)), roc_auc_score(y_te, probs)


print("=== LR Stride Sweep (AUC averaged over 3 seeds) ===\n")
lr_sweep = {}

for wk, wsize in WSIZE.items():
    lr_sweep[wk] = {}
    print(f"  {wk}")
    for stride in STRIDES[wk]:
        X_tr, y_tr, X_te, y_te = make_windows_strided(df, wsize, stride)
        Xtr_agg = aggregate_window(X_tr)
        Xte_agg = aggregate_window(X_te)

        f1s, aucs = [], []
        for s in SEEDS:
            f1, auc = _run_lr(Xtr_agg, y_tr, Xte_agg, y_te, s)
            f1s.append(f1); aucs.append(auc)

        lr_sweep[wk][stride] = {"auc_mean": np.mean(aucs), "f1_mean": np.mean(f1s)}
        label = STRIDE_LABELS[wk][stride]
        print(f"    stride={label:<12}  AUC={np.mean(aucs):.4f}  F1={np.mean(f1s):.4f}")
    print()

=== LR Stride Sweep (AUC averaged over 3 seeds) ===

  W2_1s


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


    stride=1s (base)     AUC=0.9411  F1=0.8988


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


    stride=0.5s          AUC=0.9497  F1=0.8906


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


    stride=0.25s         AUC=0.9544  F1=0.8800


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


    stride=0.125s        AUC=0.9582  F1=0.8761


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


    stride=0.1s          AUC=0.9591  F1=0.8714

  W3_5s


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


    stride=5s (base)     AUC=0.9483  F1=0.9320


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


    stride=4s            AUC=0.9425  F1=0.9283


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


    stride=2s            AUC=0.9418  F1=0.9200


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


    stride=1s            AUC=0.9445  F1=0.9109


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


    stride=0.5s          AUC=0.9459  F1=0.8990

  W4_10s


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


    stride=10s (base)    AUC=0.9505  F1=0.9510


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


    stride=7s            AUC=0.9549  F1=0.9510


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


    stride=4s            AUC=0.9497  F1=0.9583


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


    stride=2s            AUC=0.9492  F1=0.9262


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


    stride=1s            AUC=0.9492  F1=0.9020
    stride=0.5s          AUC=0.9492  F1=0.8790



/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


The sweep reveals three distinct patterns:

- **W2 (1 s):** AUC rises monotonically as stride shrinks more overlap
  always helps. Best stride: **0.1 s** (8 rows, AUC 0.9591 vs 0.9411 at base).
  The F1 trend is inverted because a larger training set shifts the decision
  boundary AUC is the more reliable selector here.

- **W3 (5 s):** The base stride (5 s, no overlap) is already optimal
  (AUC 0.9483). Overlap hurts slightly the 726 non overlapping windows are
  already sufficient for LR at this window size.

- **W4 (10 s):** A modest overlap of 7 s (stride = 560 rows) gives the best
  AUC (0.9549 vs 0.9505 at base). Beyond that, gains plateau LR saturates
  quickly as window size grows.

These three best strides are used for all subsequent model evaluations.

## 5 - Best Stride Selection

In [6]:
BEST_STRIDE = {}

print("Best stride per window size (maximises LR AUC):\n")
for wk in WSIZE:
    best  = max(lr_sweep[wk], key=lambda s: lr_sweep[wk][s]["auc_mean"])
    BEST_STRIDE[wk] = best
    label = STRIDE_LABELS[wk][best]
    auc   = lr_sweep[wk][best]["auc_mean"]
    f1    = lr_sweep[wk][best]["f1_mean"]
    print(f"  {wk}  =>  stride={label}  AUC={auc:.4f}  F1={f1:.4f}")

print(f"\nBEST_STRIDE = {BEST_STRIDE}")

Best stride per window size (maximises LR AUC):

  W2_1s  =>  stride=0.1s  AUC=0.9591  F1=0.8714
  W3_5s  =>  stride=5s (base)  AUC=0.9483  F1=0.9320
  W4_10s  =>  stride=7s  AUC=0.9549  F1=0.9510

BEST_STRIDE = {'W2_1s': 8, 'W3_5s': 400, 'W4_10s': 560}


| Window | Best stride | AUC | F1 |
|--------|-------------|-----|----|
| W2_1s  | 0.1 s (8 rows) | 0.9591 | 0.8714 |
| W3_5s  | 5 s base (400 rows) | 0.9483 | 0.9320 |
| W4_10s | 7 s (560 rows) | 0.9549 | 0.9510 |

`BEST_STRIDE` is now fixed and used unchanged for every model below.
Note that W3 keeps the non overlapping stride overlap does not always win.

## 6 - Logistic Regression, Final Evaluation

In [7]:
print("=== LR @ best stride ===\n")
for wk, wsize in WSIZE.items():
    stride   = BEST_STRIDE[wk]
    X_tr, y_tr, X_te, y_te = make_windows_strided(df, wsize, stride)
    Xtr_agg  = aggregate_window(X_tr)
    Xte_agg  = aggregate_window(X_te)

    f1s, aucs, probs_list = [], [], []
    for s in SEEDS:
        scaler = StandardScaler()
        Xtr_sc = scaler.fit_transform(Xtr_agg)
        Xte_sc = scaler.transform(Xte_agg)
        clf    = LogisticRegression(max_iter=1000, C=1.0,
                                    class_weight="balanced",
                                    random_state=s, n_jobs=-1)
        clf.fit(Xtr_sc, y_tr)
        probs  = clf.predict_proba(Xte_sc)[:, 1]
        f1s.append(f1_score(y_te, clf.predict(Xte_sc)))
        aucs.append(roc_auc_score(y_te, probs))
        probs_list.append(probs)

    f1_mean, auc_mean = np.mean(f1s), np.mean(aucs)
    avg_probs = np.mean(probs_list, axis=0)

    results[wk]["LR"] = {
        "f1": f1_mean, "auc": auc_mean,
        "y_te": y_te, "y_score": avg_probs,
        "y_pred": (avg_probs >= 0.5).astype(int),
    }
    label = STRIDE_LABELS[wk][stride]
    print(f"  {wk}  stride={label}")
    print(f"    F1={f1_mean:.4f}  AUC={auc_mean:.4f}")
    print(f"    Δ vs baseline:  ΔF1={f1_mean - BASELINE['LR'][wk]['f1']:+.4f}"
          f"  ΔAUC={auc_mean - BASELINE['LR'][wk]['auc']:+.4f}\n")

=== LR @ best stride ===



/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


  W2_1s  stride=0.1s
    F1=0.8714  AUC=0.9591
    Δ vs baseline:  ΔF1=+0.0155  ΔAUC=+0.0484



/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


  W3_5s  stride=5s (base)
    F1=0.9320  AUC=0.9483
    Δ vs baseline:  ΔF1=+0.1290  ΔAUC=+0.1334

  W4_10s  stride=7s
    F1=0.9510  AUC=0.9549
    Δ vs baseline:  ΔF1=+0.1603  ΔAUC=+0.1511



/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1435: FutureWarning: 'n_jobs' has no effect since 1.8 and will be removed in 1.10. You provided 'n_jobs=-1', please leave it unspecified.
  warnings.warn(msg, category=FutureWarning)


| Window | Stride | F1 | AUC | ΔF1 | ΔAUC |
|--------|--------|----|-----|-----|------|
| W2_1s  | 0.1 s  | 0.8714 | 0.9591 | +0.016 | +0.048 |
| W3_5s  | 5 s (base) | 0.9320 | 0.9483 | +0.129 | +0.133 |
| W4_10s | 7 s    | 0.9510 | 0.9549 | +0.160 | +0.151 |

LR benefits most at larger windows the gains at W3 and W4 are substantial
(+0.129 and +0.160 F1). This confirms that the baseline LR was not a weak
model it was simply **starved of training samples**. At W4 with only 360
non overlapping windows, the decision boundary was underfitted 7 s stride
expands the training set to 512 windows and closes most of the gap.

W2 gains are modest (+0.016 F1) because even the non overlapping baseline
already had 3,650 windows enough for LR to fit a good boundary.

## 7 - One Class SVM, Final Evaluation

In [8]:
print("=== OCSVM @ best stride ===\n")
for wk, wsize in WSIZE.items():
    stride   = BEST_STRIDE[wk]
    X_tr, y_tr, X_te, y_te = make_windows_strided(df, wsize, stride)
    Xtr_agg  = aggregate_window(X_tr)
    Xte_agg  = aggregate_window(X_te)

    f1s, aucs, scores_list, preds_list = [], [], [], []
    for s in SEEDS:
        rng       = np.random.RandomState(s)
        scaler    = StandardScaler()
        Xtr_sc    = scaler.fit_transform(Xtr_agg)
        Xte_sc    = scaler.transform(Xte_agg)
        Xtr_empty = Xtr_sc[y_tr == 0]

        if len(Xtr_empty) > 20_000:
            idx = rng.choice(len(Xtr_empty), 20_000, replace=False)
            Xtr_empty = Xtr_empty[idx]

        ocsvm  = OneClassSVM(kernel="rbf", nu=0.1, gamma="scale")
        ocsvm.fit(Xtr_empty)
        scores = -ocsvm.decision_function(Xte_sc)  # higher = more anomalous
        preds  = (ocsvm.predict(Xte_sc) == -1).astype(int)  # -1 = outlier = occupied
        f1s.append(f1_score(y_te, preds))
        aucs.append(roc_auc_score(y_te, scores))
        scores_list.append(scores); preds_list.append(preds)

    f1_mean, auc_mean = np.mean(f1s), np.mean(aucs)
    avg_scores = np.mean(scores_list, axis=0)
    y_pred_mv  = (np.mean(preds_list, axis=0) >= 0.5).astype(int)  # majority vote

    results[wk]["OCSVM"] = {
        "f1": f1_mean, "auc": auc_mean,
        "y_te": y_te, "y_score": avg_scores, "y_pred": y_pred_mv,
    }
    label = STRIDE_LABELS[wk][stride]
    print(f"  {wk}  stride={label}")
    print(f"    F1={f1_mean:.4f}  AUC={auc_mean:.4f}")
    print(f"    Δ vs baseline:  ΔF1={f1_mean - BASELINE['OCSVM'][wk]['f1']:+.4f}"
          f"  ΔAUC={auc_mean - BASELINE['OCSVM'][wk]['auc']:+.4f}\n")

=== OCSVM @ best stride ===

  W2_1s  stride=0.1s
    F1=0.9299  AUC=0.9516
    Δ vs baseline:  ΔF1=+0.1480  ΔAUC=+0.0332

  W3_5s  stride=5s (base)
    F1=0.9121  AUC=0.9694
    Δ vs baseline:  ΔF1=-0.0287  ΔAUC=-0.0069

  W4_10s  stride=7s
    F1=0.8931  AUC=0.9852
    Δ vs baseline:  ΔF1=-0.0605  ΔAUC=-0.0083



| Window | Stride | F1 | AUC | ΔF1 | ΔAUC |
|--------|--------|----|-----|-----|------|
| W2_1s  | 0.1 s  | 0.9299 | 0.9516 | +0.148 | +0.033 |
| W3_5s  | 5 s (base) | 0.9121 | 0.9694 | −0.029 | −0.007 |
| W4_10s | 7 s    | 0.8931 | 0.9852 | −0.061 | −0.008 |

OCSVM is the only model that **degrades** at W3 and W4 with overlapping
windows. The reason is structural OCSVM is trained exclusively on the
*empty* class and fits a tight hypersphere around it. At W3/W4 the baseline
was already near-perfect (F1 0.94/0.95), so the training set was not the
bottleneck. Overlapping instead introduces **near duplicate windows** each
new window shares 7–9 s of frames with its neighbour which artificially
inflates the density of the empty cluster without adding new information.
This crowds the one class boundary and shifts it in ways that hurt recall on
occupied windows.

W2 is the exception: the baseline OCSVM was genuinely undersupplied at 1 s
windows, so the extra samples from 0.1 s stride deliver a real +0.148 F1 gain.

## 8 - Convolutional Autoencoder, Final Evaluation

In [ ]:
class Autoencoder(nn.Module):
    def __init__(self):
        super().__init__()
        self.encoder = nn.Sequential(
            nn.Linear(110, 64), nn.ReLU(),
            nn.Linear(64,  16), nn.ReLU(),
        )
        self.decoder = nn.Sequential(
            nn.Linear(16,  64), nn.ReLU(),
            nn.Linear(64, 110),
        )

    def forward(self, x):
        return self.decoder(self.encoder(x))


def _recon_errors(model, X_scaled):
    """Per-sample MSE reconstruction error (numpy in, numpy out)"""
    model.eval()
    with torch.no_grad():
        X_t   = torch.tensor(X_scaled, dtype=torch.float32).to(DEVICE)
        recon = model(X_t).cpu().numpy()
    return ((X_scaled - recon) ** 2).mean(axis=1)


def _train_ae(X_empty):
    """Train AE on a scaled empty window array return trained model"""
    torch.manual_seed(SEED)
    loader  = DataLoader(TensorDataset(torch.tensor(X_empty)),
                         batch_size=BATCH_AE, shuffle=True)
    model   = Autoencoder().to(DEVICE)
    opt     = torch.optim.AdamW(model.parameters(), lr=LR_AE)
    loss_fn = nn.MSELoss()
    model.train()
    for epoch in range(1, EPOCHS_AE + 1):
        for (xb,) in loader:
            xb = xb.to(DEVICE)
            opt.zero_grad(); loss_fn(model(xb), xb).backward(); opt.step()
        if epoch % 10 == 0:
            print(f"    epoch {epoch}/{EPOCHS_AE}")
    return model


print("=== Autoencoder (val split Youden-J threshold) ===\n")
for wk, wsize in WSIZE.items():
    stride   = BEST_STRIDE[wk]
    X_tr, y_tr, X_te, y_te = make_windows_strided(df, wsize, stride)

    Xtr_agg = aggregate_window(X_tr).astype(np.float32)
    Xte_agg = aggregate_window(X_te).astype(np.float32)

    scaler      = StandardScaler()
    Xtr_sc      = scaler.fit_transform(Xtr_agg)
    Xte_sc      = scaler.transform(Xte_agg)

    # stratified val split for threshold calibration
    Xtr_sub, X_val, ytr_sub, y_val = train_test_split(
        Xtr_sc, y_tr, test_size=AE_VAL_FRAC, random_state=SEED, stratify=y_tr,
    )
    Xtr_empty = Xtr_sub[ytr_sub == 0]

    print(f"  {wk}  — AE train empty: {len(Xtr_empty):,}  val: {len(X_val):,}")
    ae_model = _train_ae(Xtr_empty)

    # Youden's J: maximise TPR - FPR on val
    val_errs              = _recon_errors(ae_model, X_val)
    fpr_v, tpr_v, thr_v  = roc_curve(y_val, val_errs)
    best_thresh           = float(thr_v[np.argmax(tpr_v - fpr_v)])

    te_errs = _recon_errors(ae_model, Xte_sc)
    y_pred  = (te_errs > best_thresh).astype(int)
    f1      = f1_score(y_te, y_pred)
    auc     = roc_auc_score(y_te, te_errs)

    results[wk]["AE"] = {
        "f1": f1, "auc": auc, "threshold": best_thresh,
        "y_te": y_te, "y_pred": y_pred, "y_score": te_errs,
    }
    print(f"  threshold={best_thresh:.5f}  F1={f1:.4f}  AUC={auc:.4f}")
    print(f"  Δ vs baseline:  ΔF1={f1 - BASELINE['AE'][wk]['f1']:+.4f}"
          f"  ΔAUC={auc - BASELINE['AE'][wk]['auc']:+.4f}\n")

=== Autoencoder (val-split Youden-J threshold) ===

  W2_1s  — AE train empty: 17,394  val: 5,469
    epoch 10/60
    epoch 20/60
    epoch 30/60
    epoch 40/60
    epoch 50/60
    epoch 60/60
  threshold=0.01606  F1=0.9725  AUC=0.9835
  Δ vs baseline:  ΔF1=+0.0215  ΔAUC=+0.0014

  W3_5s  — AE train empty: 346  val: 109
    epoch 10/60
    epoch 20/60
    epoch 30/60
    epoch 40/60
    epoch 50/60
    epoch 60/60
  threshold=0.15488  F1=0.9304  AUC=0.9454
  Δ vs baseline:  ΔF1=+0.0045  ΔAUC=-0.0395

  W4_10s  — AE train empty: 243  val: 77
    epoch 10/60
    epoch 20/60
    epoch 30/60
    epoch 40/60
    epoch 50/60
    epoch 60/60
  threshold=0.17948  F1=0.9489  AUC=0.9536
  Δ vs baseline:  ΔF1=+0.0060  ΔAUC=-0.0026



| Window | Stride | Threshold | F1 | AUC | ΔF1 | ΔAUC |
|--------|--------|-----------|-----|-----|-----|------|
| W2_1s  | 0.1 s  | 0.01606 | 0.9725 | 0.9835 | +0.022 | +0.001 |
| W3_5s  | 5 s (base) | 0.15488 | 0.9304 | 0.9454 | +0.005 | −0.040 |
| W4_10s | 7 s    | 0.17948 | 0.9489 | 0.9536 | +0.006 | −0.003 |

The Autoencoder was already the strongest baseline model at W2 (F1 0.951),
so gains are modest. W2 sees a clean +0.022 F1 improvement the larger
training set helps the encoder learn a tighter reconstruction of empty frames.

At W3 and W4 the F1 improves slightly but **AUC drops** (−0.040 at W3).
The likely cause is threshold sensitivity: Youden's J is calibrated on a
validation set split from the overlapping training windows. Near-duplicate
windows in the val set make the threshold slightly overfit the model ranks
windows well (AUC) but the operating point drifts. The effect is small in
absolute terms and F1 still improves, so overlapping is net positive even
for the AE.

The threshold rises with window size (0.016 => 0.155 => 0.179) because longer
windows produce higher mean reconstruction errors the aggregated 110-feature
vector captures more variance at larger temporal scales.

## 9 - 1D CNN, Final Evaluation

In [10]:
class CNN1D(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv_block = nn.Sequential(
            nn.Conv1d(N_FEATURES, 64,  kernel_size=3, padding=1), nn.ReLU(),
            nn.MaxPool1d(2),
            nn.Conv1d(64,         128, kernel_size=3, padding=1), nn.ReLU(),
            nn.AdaptiveAvgPool1d(1),
        )
        self.classifier = nn.Sequential(
            nn.Flatten(),
            nn.Linear(128, 64), nn.ReLU(), nn.Dropout(0.3),
            nn.Linear(64, 1),
        )

    def forward(self, x):           # x: (B, T, F)
        return self.classifier(self.conv_block(x.permute(0, 2, 1)))


def _train_cnn(X_tr, y_tr, wk):
    torch.manual_seed(SEED)
    # stratified val split
    Xtr_t, Xval_t, ytr_t, yval_t = train_test_split(
        X_tr, y_tr, test_size=0.15, random_state=SEED, stratify=y_tr,
    )
    bs = BATCH_CNN[wk]
    loader_tr = DataLoader(
        TensorDataset(torch.tensor(Xtr_t),
                      torch.tensor(ytr_t, dtype=torch.float32)),
        batch_size=bs, shuffle=True)
    loader_vl = DataLoader(
        TensorDataset(torch.tensor(Xval_t),
                      torch.tensor(yval_t, dtype=torch.float32)),
        batch_size=bs * 2)

    model   = CNN1D().to(DEVICE)
    opt     = torch.optim.AdamW(model.parameters(), lr=LR_CNN, weight_decay=1e-4)
    sched   = torch.optim.lr_scheduler.CosineAnnealingLR(opt, T_max=EPOCHS_CNN, eta_min=1e-5)
    loss_fn = nn.BCEWithLogitsLoss()

    best_val, patience_ctr, best_state = float("inf"), 0, None
    for epoch in range(1, EPOCHS_CNN + 1):
        model.train()
        for Xb, yb in loader_tr:
            Xb, yb = Xb.to(DEVICE), yb.to(DEVICE)
            opt.zero_grad()
            loss_fn(model(Xb).squeeze(1), yb).backward()
            opt.step()
        sched.step()

        model.eval()
        val_loss = 0.0
        with torch.no_grad():
            for Xb, yb in loader_vl:
                Xb, yb = Xb.to(DEVICE), yb.to(DEVICE)
                val_loss += loss_fn(model(Xb).squeeze(1), yb).item() * len(Xb)
        val_loss /= len(Xval_t)

        if val_loss < best_val - 1e-5:
            best_val = val_loss; patience_ctr = 0
            best_state = {k: v.cpu().clone() for k, v in model.state_dict().items()}
        else:
            patience_ctr += 1

        if epoch % 10 == 0:
            print(f"    epoch {epoch:>3}/{EPOCHS_CNN}  val_loss={val_loss:.4f}"
                  f"  lr={sched.get_last_lr()[0]:.5f}")
        if patience_ctr >= CNN_PATIENCE:
            print(f"    => early stop at epoch {epoch}"); break

    model.load_state_dict({k: v.to(DEVICE) for k, v in best_state.items()})
    return model


print("=== CNN (CosineAnnealingLR + early stopping) ===\n")
for wk, wsize in WSIZE.items():
    stride   = BEST_STRIDE[wk]
    X_tr, y_tr, X_te, y_te = make_windows_strided(df, wsize, stride)

    # scale feature-wise across all time steps
    scaler  = StandardScaler()
    scaler.fit(X_tr.reshape(-1, N_FEATURES))
    X_tr_sc = scaler.transform(X_tr.reshape(-1, N_FEATURES)).reshape(X_tr.shape)
    X_te_sc = scaler.transform(X_te.reshape(-1, N_FEATURES)).reshape(X_te.shape)

    print(f"  {wk}  (train={len(X_tr):,})")
    model = _train_cnn(X_tr_sc, y_tr, wk)

    model.eval()
    with torch.no_grad():
        logits = model(torch.tensor(X_te_sc).to(DEVICE)).squeeze(1).cpu().numpy()
    probs  = torch.sigmoid(torch.tensor(logits)).numpy()
    y_pred = (probs >= 0.5).astype(int)
    f1     = f1_score(y_te, y_pred)
    auc    = roc_auc_score(y_te, probs)

    results[wk]["CNN"] = {
        "f1": f1, "auc": auc,
        "y_te": y_te, "y_pred": y_pred, "y_score": probs,
    }
    label = STRIDE_LABELS[wk][stride]
    print(f"  stride={label}  F1={f1:.4f}  AUC={auc:.4f}")
    print(f"  Δ vs baseline:  ΔF1={f1 - BASELINE['CNN'][wk]['f1']:+.4f}"
          f"  ΔAUC={auc - BASELINE['CNN'][wk]['auc']:+.4f}\n")

=== CNN (CosineAnnealingLR + early stopping) ===

  W2_1s  (train=36,460)
    epoch  10/80  val_loss=0.0018  lr=0.00096
    epoch  20/80  val_loss=0.0045  lr=0.00086
    epoch  30/80  val_loss=0.0000  lr=0.00069
    => early stop at epoch 33
  stride=0.1s  F1=0.9612  AUC=0.9857
  Δ vs baseline:  ΔF1=+0.1356  ΔAUC=+0.0172

  W3_5s  (train=726)
    epoch  10/80  val_loss=0.0064  lr=0.00096
    epoch  20/80  val_loss=0.0006  lr=0.00086
    epoch  30/80  val_loss=0.0003  lr=0.00069
    epoch  40/80  val_loss=0.0003  lr=0.00051
    epoch  50/80  val_loss=0.0001  lr=0.00032
    epoch  60/80  val_loss=0.0001  lr=0.00015
    => early stop at epoch 65
  stride=5s (base)  F1=0.8280  AUC=0.9653
  Δ vs baseline:  ΔF1=+0.0166  ΔAUC=+0.0110

  W4_10s  (train=512)
    epoch  10/80  val_loss=0.0043  lr=0.00096
    epoch  20/80  val_loss=0.0010  lr=0.00086
    epoch  30/80  val_loss=0.0004  lr=0.00069
    epoch  40/80  val_loss=0.0002  lr=0.00051
    epoch  50/80  val_loss=0.0001  lr=0.00032
    epoch 

| Window | Stride | Stopped at | F1 | AUC | ΔF1 | ΔAUC |
|--------|--------|------------|-----|-----|-----|------|
| W2_1s  | 0.1 s  | epoch 33 | 0.9612 | 0.9857 | +0.136 | +0.017 |
| W3_5s  | 5 s (base) | epoch 65 | 0.8280 | 0.9653 | +0.017 | +0.011 |
| W4_10s | 7 s    | epoch 67 | 0.8235 | 0.9679 | +0.014 | +0.035 |

The CNN produces the **single largest absolute gain** in this notebook:
W2 F1 jumps from 0.826 to **0.961** (+0.136). The explanation is visible in
the early-stopping epoch: W2 converges at epoch 33 on 36,460 training
windows the model finds a strong solution quickly because the data is
abundant. The baseline CNN at W2 had only 3,650 windows, which was
insufficient for the convolutional layers to learn reliable temporal patterns
in a 1 s window.

W3 and W4 stop much later (epochs 65 and 67) despite having far fewer
training windows (726 and 512). The val loss plateaus very slowly the
model is learning on a small dataset and the CosineAnnealingLR schedule
keeps the learning rate high enough that it never fully settles. The gains
are modest (+0.014–+0.017 F1) but AUC improves more meaningfully at W4
(+0.035), suggesting the model ranks windows better even when the F1
threshold is imprecise.

## 10 - Transformer, Final Evaluation

In [ ]:
class PositionalEncoding(nn.Module):
    """Sinusoidal positional encoding supports sequences up to max_len"""
    def __init__(self, d_model, max_len=1200, dropout=0.1):
        super().__init__()
        self.dropout = nn.Dropout(dropout)
        pe  = torch.zeros(max_len, d_model)
        pos = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
        div = torch.exp(
            torch.arange(0, d_model, 2, dtype=torch.float) * (-math.log(10_000.0) / d_model)
        )
        pe[:, 0::2] = torch.sin(pos * div)
        pe[:, 1::2] = torch.cos(pos * div)
        self.register_buffer("pe", pe.unsqueeze(0))   # (1, max_len, d_model)

    def forward(self, x):
        return self.dropout(x + self.pe[:, :x.size(1)])


class TransformerClassifier(nn.Module):
    """CLS token Transformer encoder for binary classification"""
    def __init__(self, n_feat=N_FEATURES, d_model=64, n_heads=4,
                 n_layers=2, dim_ff=128, dropout=0.1):
        super().__init__()
        self.input_proj = nn.Linear(n_feat, d_model)
        self.pos_enc    = PositionalEncoding(d_model, dropout=dropout)
        self.cls_token  = nn.Parameter(torch.zeros(1, 1, d_model))
        enc_layer       = nn.TransformerEncoderLayer(
            d_model=d_model, nhead=n_heads,
            dim_feedforward=dim_ff, dropout=dropout, batch_first=True,
        )
        self.encoder    = nn.TransformerEncoder(enc_layer, num_layers=n_layers)
        self.classifier = nn.Sequential(
            nn.Linear(d_model, 32), nn.ReLU(),
            nn.Linear(32, 1),
        )

    def forward(self, x):                              # x: (B, T, F)
        x   = self.pos_enc(self.input_proj(x))        # (B, T, d_model)
        cls = self.cls_token.expand(x.size(0), -1, -1)
        x   = self.encoder(torch.cat([cls, x], dim=1))
        return self.classifier(x[:, 0])               # CLS output


def _train_transformer(X_tr, y_tr, wk):
    torch.manual_seed(SEED)
    Xtr_t, Xval_t, ytr_t, yval_t = train_test_split(
        X_tr, y_tr, test_size=0.15, random_state=SEED, stratify=y_tr,
    )
    bs = BATCH_TR[wk]
    loader_tr = DataLoader(
        TensorDataset(torch.tensor(Xtr_t),
                      torch.tensor(ytr_t, dtype=torch.float32)),
        batch_size=bs, shuffle=True)
    loader_vl = DataLoader(
        TensorDataset(torch.tensor(Xval_t),
                      torch.tensor(yval_t, dtype=torch.float32)),
        batch_size=bs * 2)

    model   = TransformerClassifier().to(DEVICE)   # always on DEVICE (CUDA on vast.ai)
    opt     = torch.optim.AdamW(model.parameters(), lr=LR_TR, weight_decay=1e-4)
    sched   = torch.optim.lr_scheduler.CosineAnnealingLR(opt, T_max=EPOCHS_TR, eta_min=1e-5)
    loss_fn = nn.BCEWithLogitsLoss()

    best_val, patience_ctr, best_state = float("inf"), 0, None
    for epoch in range(1, EPOCHS_TR + 1):
        model.train()
        for Xb, yb in loader_tr:
            Xb, yb = Xb.to(DEVICE), yb.to(DEVICE)
            opt.zero_grad()
            loss_fn(model(Xb).squeeze(1), yb).backward()
            opt.step()
        sched.step()

        model.eval()
        val_loss = 0.0
        with torch.no_grad():
            for Xb, yb in loader_vl:
                Xb, yb = Xb.to(DEVICE), yb.to(DEVICE)
                val_loss += loss_fn(model(Xb).squeeze(1), yb).item() * len(Xb)
        val_loss /= len(Xval_t)

        if val_loss < best_val - 1e-5:
            best_val = val_loss; patience_ctr = 0
            best_state = {k: v.cpu().clone() for k, v in model.state_dict().items()}
        else:
            patience_ctr += 1

        if epoch % 10 == 0:
            print(f"    epoch {epoch:>3}/{EPOCHS_TR}  val_loss={val_loss:.4f}"
                  f"  lr={sched.get_last_lr()[0]:.5f}")
        if patience_ctr >= TR_PATIENCE:
            print(f"    => early stop at epoch {epoch}"); break

    model.load_state_dict({k: v.to(DEVICE) for k, v in best_state.items()})
    return model


print("=== Transformer (CosineAnnealingLR + early stopping) ===\n")
for wk, wsize in WSIZE.items():
    stride   = BEST_STRIDE[wk]
    X_tr, y_tr, X_te, y_te = make_windows_strided(df, wsize, stride)

    scaler  = StandardScaler()
    scaler.fit(X_tr.reshape(-1, N_FEATURES))
    X_tr_sc = scaler.transform(X_tr.reshape(-1, N_FEATURES)).reshape(X_tr.shape)
    X_te_sc = scaler.transform(X_te.reshape(-1, N_FEATURES)).reshape(X_te.shape)

    print(f"  {wk}  (train={len(X_tr):,})")
    model = _train_transformer(X_tr_sc, y_tr, wk)

    model.eval()
    with torch.no_grad():
        logits = model(torch.tensor(X_te_sc).to(DEVICE)).squeeze(1).cpu().numpy()
    probs  = torch.sigmoid(torch.tensor(logits)).numpy()
    y_pred = (probs >= 0.5).astype(int)
    f1     = f1_score(y_te, y_pred)
    auc    = roc_auc_score(y_te, probs)

    results[wk]["Transformer"] = {
        "f1": f1, "auc": auc,
        "y_te": y_te, "y_pred": y_pred, "y_score": probs,
    }
    label = STRIDE_LABELS[wk][stride]
    print(f"  stride={label}  F1={f1:.4f}  AUC={auc:.4f}")
    print(f"  Δ vs baseline:  ΔF1={f1 - BASELINE['Transformer'][wk]['f1']:+.4f}"
          f"  ΔAUC={auc - BASELINE['Transformer'][wk]['auc']:+.4f}\n")

=== Transformer (CosineAnnealingLR + early stopping) ===

  W2_1s  (train=36,460)
    epoch  10/100  val_loss=0.0027  lr=0.00098
    epoch  20/100  val_loss=0.0073  lr=0.00091
    epoch  30/100  val_loss=0.0001  lr=0.00080
    epoch  40/100  val_loss=0.0002  lr=0.00066
    epoch  50/100  val_loss=0.0000  lr=0.00051
    epoch  60/100  val_loss=0.0004  lr=0.00035
    => early stop at epoch 68
  stride=0.1s  F1=0.9572  AUC=0.9825
  Δ vs baseline:  ΔF1=+0.0440  ΔAUC=+0.0254

  W3_5s  (train=726)
    epoch  10/100  val_loss=0.0049  lr=0.00098
    epoch  20/100  val_loss=0.0007  lr=0.00091
    epoch  30/100  val_loss=0.0471  lr=0.00080
    epoch  40/100  val_loss=0.0003  lr=0.00066
    epoch  50/100  val_loss=0.0002  lr=0.00051
    epoch  60/100  val_loss=0.0002  lr=0.00035
    epoch  70/100  val_loss=0.0001  lr=0.00021
    epoch  80/100  val_loss=0.0001  lr=0.00010
    epoch  90/100  val_loss=0.0001  lr=0.00003
    epoch 100/100  val_loss=0.0001  lr=0.00001
    => early stop at epoch 100
  

| Window | Stride | Stopped at | F1 | AUC | ΔF1 | ΔAUC |
|--------|--------|------------|-----|-----|-----|------|
| W2_1s  | 0.1 s  | epoch 68 | 0.9572 | 0.9825 | +0.044 | +0.025 |
| W3_5s  | 5 s (base) | epoch 100 | 0.8211 | 0.9482 | +0.015 | +0.033 |
| W4_10s | 7 s    | epoch 46 | 0.7977 | 0.9457 | −0.000 | +0.004 |

W2 shows a solid +0.044 F1 gain the Transformer benefits from 36,460
training windows to learn the self attention patterns that distinguish
empty and occupied CSI signals at 1 s resolution.

W3 runs to the full 100 epochs without triggering early stopping, meaning
the model never plateaued it kept slowly improving on 726 windows right
up to the budget limit. This suggests more epochs could squeeze out further
gains, though the improvement would likely be marginal given the small
training set.

W4 is the most striking result: **ΔF1 = 0.000** exactly. The Transformer
at 10 s windows with 512 training samples hits an early stop at epoch 46
the val loss spikes at epoch 40 (0.0026, up from 0.0006) before the model
cannot recover within the patience window. The ranking improves slightly
(ΔAUC +0.004) but the decision boundary is identical to the non-overlapping
baseline. A 10 s context window may already capture the full temporal
structure the Transformer needs, making additional overlapping windows
redundant rather than informative.

## 11 - Cross Model Comparison

In [ ]:
MODELS = ["LR", "OCSVM", "AE", "CNN", "Transformer"]

print("=" * 72)
print("SUMMARY Overlapping Stride vs Non Overlapping Baseline")
print("=" * 72)
print(f"\n{'Model':<14} {'Window':<10}  {'Base F1':>7}  {'Ovlp F1':>7}  {'ΔF1':>6}  "
      f"{'Base AUC':>8}  {'Ovlp AUC':>8}  {'ΔAUC':>6}")
print("-" * 72)

for m in MODELS:
    for wk in WSIZE:
        if m not in results[wk]:
            continue
        f1  = results[wk][m]["f1"]
        auc = results[wk][m]["auc"]
        bf1 = BASELINE[m][wk]["f1"]
        ba  = BASELINE[m][wk]["auc"]
        print(f"  {m:<12} {wk:<10}  {bf1:>7.4f}  {f1:>7.4f}  {f1-bf1:>+6.4f}  "
              f"{ba:>8.4f}  {auc:>8.4f}  {auc-ba:>+6.4f}")
    print()

colours = {
    ("Logistic Regression", "#2C7BB6", "--"),
    ("1D CNN",              "#D97B2B", "--"),
    ("Transformer",         "#4DAF4A", "--"),
    ("One-Class SVM",       "#984EA3", "-"),
    ("Conv Autoencoder",    "#E41A1C", "-"),
}

# --- ROC curves ---------------------------------------------------------------
fig, axes = plt.subplots(1, 3, figsize=(15, 5))

for ax, wk in zip(axes, WSIZE):
    for m in MODELS:
        if m not in results[wk]:
            continue
        r   = results[wk][m]
        fpr, tpr, _ = roc_curve(r["y_te"], r["y_score"])
        ax.plot(fpr, tpr, label=f"{m} (AUC={r['auc']:.3f})",
                color=colours[m], lw=1.8)
    ax.plot([0, 1], [0, 1], "k--", lw=0.8)
    ax.set_title(f"{wk} overlapping stride", fontsize=11)
    ax.set_xlabel("FPR"); ax.set_ylabel("TPR")
    ax.legend(fontsize=8); ax.grid(alpha=0.3)

plt.suptitle("ROC Curve — Best Overlapping Stride (all models)", fontsize=13, y=1.02)
plt.tight_layout()
plt.savefig(f'{SAVE_DIR}/06_roc_overlapping.png', dpi=150, bbox_inches='tight')
plt.show()

# --- ΔF1 bar chart ------------------------------------------------------------
fig2, axes2 = plt.subplots(1, 3, figsize=(15, 4), sharey=True)
x = np.arange(len(MODELS))

for ax, wk in zip(axes2, WSIZE):
    deltas = []
    for m in MODELS:
        if m in results[wk]:
            deltas.append(results[wk][m]["f1"] - BASELINE[m][wk]["f1"])
        else:
            deltas.append(0.0)
    bars = ax.bar(x, deltas, color=[colours[m] for m in MODELS], alpha=0.85)
    ax.axhline(0, color="black", lw=0.8)
    ax.set_xticks(x); ax.set_xticklabels(MODELS, rotation=25, ha="right", fontsize=9)
    ax.set_title(f"{wk}", fontsize=11); ax.set_ylabel("ΔF1 vs baseline")
    ax.grid(axis="y", alpha=0.3)
    for bar, d in zip(bars, deltas):
        ax.text(bar.get_x() + bar.get_width() / 2, bar.get_height() + 0.002,
                f"{d:+.3f}", ha="center", va="bottom", fontsize=8)

plt.suptitle("F1 gain from overlapping stride vs non-overlapping baseline", fontsize=12)
plt.tight_layout()
plt.savefig(f'{SAVE_DIR}/06_delta_f1_overlapping.png', dpi=150, bbox_inches='tight')
plt.show()

SUMMARY Overlapping Stride vs Non-Overlapping Baseline

Model          Window      Base F1  Ovlp F1     ΔF1  Base AUC  Ovlp AUC    ΔAUC
------------------------------------------------------------------------
  LR           W2_1s        0.8559   0.8714  +0.0155    0.9107    0.9591  +0.0484
  LR           W3_5s        0.8030   0.9320  +0.1290    0.8149    0.9483  +0.1334
  LR           W4_10s       0.7907   0.9510  +0.1603    0.8038    0.9549  +0.1511

  OCSVM        W2_1s        0.7819   0.9299  +0.1480    0.9184    0.9516  +0.0332
  OCSVM        W3_5s        0.9408   0.9121  -0.0287    0.9763    0.9694  -0.0069
  OCSVM        W4_10s       0.9536   0.8931  -0.0605    0.9935    0.9852  -0.0083

  AE           W2_1s        0.9510   0.9725  +0.0215    0.9821    0.9835  +0.0014
  AE           W3_5s        0.9259   0.9304  +0.0045    0.9849    0.9454  -0.0395
  AE           W4_10s       0.9429   0.9489  +0.0060    0.9562    0.9536  -0.0026

  CNN          W2_1s        0.8256   0.9612  +0.13

![ROC curves at best overlapping stride](full-study-outputs/06_roc_overlapping.png)

![ΔF1 gain from overlapping stride vs non overlapping baseline](full-study-outputs/06_delta_f1_overlapping.png)

The summary table and plots reveal four conclusions:

**1. Short windows benefit most universally.**
Every model improves at W2 (1 s). At 1 s, even the non overlapping baseline
had enough windows for statistical models (LR, OCSVM) but not for neural ones.
Overlapping fixes both it multiplies samples for neural models *and* shifts
the LR/OCSVM decision boundary with more varied examples.

**2. LR was sample starved at large windows.**
The +0.160 F1 gain at W4 is the largest in the table and it comes from the
simplest model. This is not a flaw in LR; it is a data problem. With only 360
non-overlapping windows, no linear classifier can fit a reliable boundary.
Overlapping recovers what the model was always capable of.

**3. OCSVM is the exception.**
One class methods are sensitive to near duplicate training points. At W3/W4,
where the baseline was already strong (F1 0.94/0.95), overlapping introduces
correlated windows that distort the empty class hypersphere without adding new
information. The degradation is small in AUC (−0.007/−0.008) but more visible
in F1 (−0.029/−0.061). For one class models, more data is not always better
data.

**4. Overlap is a hyperparameter, not a free improvement.**
The best stride varies by both model and window size. The optimal choice ranges
from no overlap (W3 LR, W3/W4 OCSVM) to 90 % overlap (W2 CNN, W2 Transformer).
Treating stride as a fixed design decision as most prior work does leaves
measurable performance on the table.